store/postgres/src/copy.rs

```diff
@@ -0,0 +1,540 @@
+                .map_err(|e| {
+                    constraint_violation!(
+                        "invalid {} table {} in CopyState {} (table {}): {}",
+                        kind,
+                        entity_type,
+                        dst,
+                        id,
+                        e
+                    )
+                })
+                .map(|table| table.clone())
+        }
+
+        cts::table
+            .filter(cts::dst.eq(dst_layout.site.id))
+            .select((
+                cts::id,
+                cts::entity_type,
+                cts::next_vid,
+                cts::target_vid,
+                cts::batch_size,
+            ))
+            .load::<(i32, String, i64, i64, i64)>(conn)?
+            .into_iter()
+            .map(|(id, entity_type, current_vid, target_vid, batch_size)| {
+                let entity_type = EntityType::new(entity_type);
+                let src =
+                    resolve_entity(src_layout, "source", &entity_type, dst_layout.site.id, id);
+                let dst = resolve_entity(
+                    dst_layout,
+                    "destination",
+                    &entity_type,
+                    dst_layout.site.id,
+                    id,
+                );
+                match (src, dst) {
+                    (Ok(src), Ok(dst)) => Ok(TableState {
+                        dst_site: dst_layout.site.clone(),
+                        src,
+                        dst,
+                        next_vid: current_vid,
+                        target_vid,
+                        batch_size,
+                    }),
+                    (Err(e), _) => Err(e),
+                    (_, Err(e)) => Err(e),
+                }
+            })
+            .collect()
+    }
+
+    fn record_progress(&self, conn: &PgConnection) -> Result<(), StoreError> {
+        use copy_table_state as cts;
+
+        let values = (
+            cts::next_vid.eq(self.next_vid),
+            cts::batch_size.eq(self.batch_size),
+        );
+        update(
+            cts::table
+                .filter(cts::dst.eq(self.dst_site.id))
+                .filter(cts::entity_type.eq(self.dst.object.as_str())),
+        )
+        .set(values)
+        .execute(conn)?;
+        Ok(())
+    }
+
+    fn record_finished(&self, conn: &PgConnection) -> Result<(), StoreError> {
+        use copy_table_state as cts;
+
+        update(
+            cts::table
+                .filter(cts::dst.eq(self.dst_site.id))
+                .filter(cts::entity_type.eq(self.dst.object.as_str())),
+        )
+        .set(cts::finished_at.eq(sql("now()")))
+        .execute(conn)?;
+        Ok(())
+    }
+
+    fn copy_batch(&mut self, conn: &PgConnection) -> Result<(), StoreError> {
+        let start = Instant::now();
+
+        rq::CopyEntityBatchQuery::new(
+            self.dst.as_ref(),
+            &self.src,
+            self.next_vid,
+            self.next_vid + self.batch_size,
```

Comment:
We're not limiting by `target_vid` here, I think that'd be good to make the behavior more predictable.

Reply:
Yes, that's actually a pretty serious bug since we would end up copying data we should not be copying.